crates/wasmparser/src/validator/component_types.rs

```diff
@@ -1183,34 +1455,40 @@ impl ComponentDefinedType {
         cases: impl Iterator<Item = &'a ComponentValType>,
         types: &TypeList,
         lowered_types: &mut LoweredTypes,
-    ) -> bool {
+    ) -> LowerResult {
         // Push the discriminant
-        if !lowered_types.try_push(ValType::I32) {
-            return false;
-        }
-
-        let start = lowered_types.len();
-
+        lowered_types.try_push(ValType::I32)?;
+
+        // First, simultaneously check that the `lowered_types` has capacity for
+        // the variants' types and join the `i`th type from each variant
+        // together to get the actual types we will push onto
+        // `lowered_types`. This phase ignores the expected types, if any, since
+        // we don't know what we will actually push onto `lowered_types` until
+        // after all the joining has taken place.
+        let mut joined_tys = SmallVec::<[_; MAX_FLAT_FUNC_PARAMS]>::new();
```

Comment:
To avoid adding `SmallVec` as a dependency, could this use `clone_without_expected_types` as the storage here, similar-ish to before?